extension/json/json_scan.cpp

```diff
@@ -0,0 +1,497 @@
+		auto loption = StringUtil::Lower(kv.first);
+		if (loption == "ignore_errors") {
+			result->ignore_errors = BooleanValue::Get(kv.second);
+		} else if (loption == "maximum_object_size") {
+			result->maximum_object_size = MaxValue<idx_t>(UIntegerValue::Get(kv.second), result->maximum_object_size);
+		} else if (loption == "format") {
+			auto format = StringUtil::Lower(StringValue::Get(kv.second));
+			if (format == "auto") {
+				options.format = JSONFormat::AUTO_DETECT;
+			} else if (format == "unstructured") {
+				options.format = JSONFormat::UNSTRUCTURED;
+			} else if (format == "newline_delimited") {
+				options.format = JSONFormat::NEWLINE_DELIMITED;
+			} else {
+				throw InvalidInputException("format must be one of ['auto', 'unstructured', 'newline_delimited']");
+			}
+		}
+	}
+
+	if (result->ignore_errors && options.format == JSONFormat::UNSTRUCTURED) {
+		throw InvalidInputException("Cannot ignore errors with unstructured format");
+	}
+
+	return std::move(result);
+}
+
+void JSONScanData::InitializeFilePaths(ClientContext &context, const vector<string> &patterns,
+                                       vector<string> &file_paths) {
+	auto &fs = FileSystem::GetFileSystem(context);
+	for (auto &file_pattern : patterns) {
+		auto found_files = fs.Glob(file_pattern, context);
+		if (found_files.empty()) {
+			throw IOException("No files found that match the pattern \"%s\"", file_pattern);
+		}
+		file_paths.insert(file_paths.end(), found_files.begin(), found_files.end());
+	}
+}
+
+void JSONScanData::Serialize(FieldWriter &writer) {
+	options.Serialize(writer);
+	writer.WriteList<string>(file_paths);
+	writer.WriteField<bool>(ignore_errors);
+	writer.WriteField<idx_t>(maximum_object_size);
+	writer.WriteField<bool>(return_json_strings);
+}
+
+void JSONScanData::Deserialize(FieldReader &reader) {
+	options.Deserialize(reader);
+	file_paths = reader.ReadRequiredList<string>();
+	ignore_errors = reader.ReadRequired<bool>();
+	maximum_object_size = reader.ReadRequired<idx_t>();
+	return_json_strings = reader.ReadRequired<bool>();
+}
+
+JSONScanGlobalState::JSONScanGlobalState(ClientContext &context, JSONScanData &bind_data_p)
+    : bind_data(bind_data_p), allocator(BufferManager::GetBufferManager(context).GetBufferAllocator()),
+      buffer_capacity(bind_data.maximum_object_size * 2), file_index(0),
+      system_threads(TaskScheduler::GetScheduler(context).NumberOfThreads()) {
+	json_readers.reserve(bind_data.file_paths.size());
+	for (idx_t i = 0; i < bind_data_p.file_paths.size(); i++) {
+		json_readers.push_back(make_unique<BufferedJSONReader>(context, bind_data.options, i, bind_data.file_paths[i]));
+	}
+}
+
+unique_ptr<GlobalTableFunctionState> JSONScanGlobalState::Init(ClientContext &context, TableFunctionInitInput &input) {
+	auto &bind_data = (JSONScanData &)*input.bind_data;
+	return make_unique<JSONScanGlobalState>(context, bind_data);
+}
+
+idx_t JSONScanGlobalState::MaxThreads() const {
+	return system_threads;
```

Comment:
I copied this from the CSV reader. Do we need to open the file to get the file size? I didn't see any other way.

The parquet reader opens only the first file and then estimates the total number of row groups for all files by multiplying the number of row groups in the first file by the number of files. Would this be OK here? This would be slightly problematic if we read a small file (which would be read using 1 thread) first and then a large file (which would require all threads), as we would underestimate the thread count by a lot.

Reply:
That is true - perhaps it is indeed better to always launch `n` tasks as it is unlikely to hurt much.